### ISSUE TITLE
Put out a new release

### ISSUE THREAD

**Issue by shish** [thecodingmachine/safe#499]
to-do:

- [x] get unit tests working in all currently-supported versions of PHP
- [x] get automatic updates working
- [x] check to see if any outstanding PRs are still useful, merge them if so
- [x] check to see if any bug reports are still relevant, add them as bullet points below
    - [ ] ?
- [x] check to see if any feature requests are super-duper important, implement them (if they aren't super-duper important I'd rather put out a stable release before adding new features)
    - [ ] ?
- [x] pick a version number (IIRC "dropping support for old versions of PHP" is defined as "needs a major version number bump", but I haven't checked that)
- [ ] what else?

**Comment by staabm** [comment]
I think we should put out a beta since its our first release

**Comment by staabm** [comment]
After we got a beta out we should chrck the phpstan-plugin whether it needs some love to work with the beta

**Comment by oojacoboo** [comment]
PHP 8.4 deprecations are an issue for anyone that doesn't ignore them entirely.  A 2.5.1 release that at least fixes this would be nice.

**Comment by maks-rafalko** [comment]
Commenting for visibility of a problem and to subscribe to a topic:

absence of a release produces a lot of deprecations for the projects that use this lib. For example, in Infection we had to stick to `dev-master` to avoid deprecation, but even this leads to issues, like https://github.com/infection/infection/issues/2015

so if you could do a quick release with the fixes for 8.4, that would be awesome.

**Comment by shish** [comment]
The deprecation warnings are a superficial and obvious issue, but the version incompatibilities run deeper than that. Speaking only for myself, not all the maintainers - I don't want to put out a new release which fixes a superficial issue, only to give people the impression that the project is healthy, and have them run into worse problems.

If you really hate the warnings, don't want to silence them, and are willing to deal with worse issues - you can switch from `thecodingmachine/safe` to `shish/safe` temporarily (it is marked as a replacement package, so you only need to switch in the top-level composer.json and it will propagate through your dependencies, and you should be able to switch back just as easily)

**Comment by jhdxr** [comment]
#421 may be a starting point as it's target php8.1, which is still under security fix for one year.

Personally I suggest to bump the version to 3.0 and targeting php 8.3 (8.2 is going to end active support in around 2 weeks), which gives us more flexibility to introduce new/bc break features/changes if necessary.

For people staying with 8.1/8.2, I think current version should be fine. We can still have bugfix on this branch (2.x) until the end of 8.2.

**Comment by indy2kro** [comment]
Hello, any update on a new release?

**Comment by shish** [comment]
I've been spending Christmas with my parents and their barely-functioning
internet connection; once I'm back my plan is to finish the "be compatible
with multiple versions of PHP at the same time" proof-of-concept and see if
the other maintainers are happy with that; then release that if they are.
Not sure what anybody else has been up to, our only communication is here
in the GitHub issues 👀

**Comment by shish** [comment]
Minor update: I got home, but seem to have picked up covid from the flight home, and my brain is a mixture of fuzz and delerium x__x

If anybody wants to help things move faster, I believe https://github.com/thecodingmachine/safe/pull/503 is our current best-bet for putting out a new release which supports php8.4 _and_ older versions - code reviews / suggestions / PRs to add the missing features / any kind of assistance are welcome <3

**Comment by indy2kro** [comment]
Any updates by any chance? We literally just need someone to click the "New release" button with the code that is now in main.

**Comment by shish** [comment]
The code that's now in the master branch is broken in subtle ways across different PHP versions - you'd be getting rid of the deprecation warnings, but letting yourself in for less-obvious more-serious issues. If you're ok with that tradeoff, and only plan on using php 8.4, you can use `shish/safe` (which is flagged as a drop-in replacement for `thecodingmachine/safe` so you only need to do that at the top-level, and it will automatically use that for dependencies too) - if you want code that runs correctly on php 8.1 - 8.5, #503 is the blocker (and I welcome any code reviews or comments there, or PRs wants to get it working themselves)

If one of the other maintainers wants to push "release" on the current code and take responsibility for the complaints, I won't stop them, but I'd rather not put out a release which gives people the impression that the code is healthy when it isn't 😅

**Comment by shish** [comment]
Also if somebody wants to take care of triaging existing bug reports / feature requests / PR to see which of those are still relevant, that would help speed things up too ❤ This project was abandoned for a couple of years before the new maintainers got permission to revive it, so there's a bunch of cruft ^^;

**Comment by shish** [comment]
#503 is now in a state where it appears to work correctly across multiple PHP versions in my own test cases - If you want to test it, you can add a dependency on "thecodingmachine/safe": "dev-pr503" in your composer.json (like this - https://github.com/shish/shimmie2/pull/1379/files) - if people could try that and let me know if it works for your projects, that'd be great <3

**Comment by shish** [comment]
With ~10 people asking where the new release is at, and 0 people giving any feedback on the current code, I'mma say "it works for me" and publish `v3.0.0-alpha1`

Please, somebody, anybody, help with testing 😅  Even just updating your `composer.json` with the dependency `"thecodingmachine/safe": "^3.0.0-alpha1"` and running unit tests and commenting "v3 does not break my unit tests" would be helpful ❤️

**Comment by josecl** [comment]
v3 does not break my unit tests 🚀 

Tested against php 8.3.16 and 8.4.3 in my midsize Laravel project.

**Comment by shish** [comment]
@staabm @OskarStark @silasjoisten is there anything y'all think needs to happen between now and `3.0.0`? While there are a bunch of open requests for supporting extra functions / edge cases / integrations with other tools, I think what's in the master branch now has everything that it _needs_ to have (ie, it is tested and working, without warnings or errors, on all currently-live versions of PHP) - and extra features can be added later?

If people keep testing and no major issues get discovered, I'm thinking to release a week from now? (Trying to balance "giving people time to test" and "getting something out of the door in order to help the people who are having issues in production right now")

**Comment by staabm** [comment]
@shish I think you are in the best position to judge whether its good enough.
I don't see something obvious at this point.

thank you for doing the hard work

**Comment by silasjoisten** [comment]
Are there any breaking changes we could make now? If not release it like it's hot! 🔥

**Comment by OskarStark** [comment]
Yes double checking for potential coming Bc breaks is also the only point on my side which should be double checked before releasing 3.0

Didn't check, but do we have an upgrade guide? That said, do we really need a new major?

**Comment by shish** [comment]
> Didn't check, but do we have an upgrade guide?

Not at the moment; it should be _mostly_ a drop-in replacement, except that some functions have been deleted from the Safe namespace (either because they were long-ago deleted upstream, or because they were long-ago made safe upstream) - which any kind of static analysis tool should catch easily


> That said, do we really need a new major?

I believe so - we're deleting functions which people might be referencing (ie, Safe 2.X contains wrappers for functions which are unsafe in 7.X and safe in 8.0; Safe 3.X only generates wrappers for functions which are unsafe in 8.1+)

**Comment by shish** [comment]
> But it works correctly in the current release and is broken in master.

I think it's actually _double_-broken in the current release, in a way where one bit of brokenness cancels out another bit of brokenness, and by coincidence it actually works 😆 But yes I see what you mean and I'll look into that right away (#546)

**Comment by OskarStark** [comment]
Ok let's go with a new major, but shall we create deprecations in 2.x for the deleted methods in 3 to make it more smooth or is it not worth the work?

**Comment by michalschroeder** [comment]
Hi, version `3.0.0-alpha1` tested on PHP `8.4.2` and all works as expected.

**Comment by shish** [comment]
> shall we create deprecations in 2.x for the deleted methods in 3

Doing a quick check, the functions being deleted during the Safe 2-to-3 transition are the functions which were deleted upstream during the PHP 7-to-8 transition (for example, `gzgetss`) - so anybody calling those functions today is _already_ getting a "function does not exist" error

(In a previous comment I also mentioned dropping support for "functions which were unsafe in the past but have been safe now for a long time", but it looks like all the examples I was thinking of actually became safe between 8.0 and 8.5 (eg `shuffle`) -- and we handle those cases automatically, so there's nothing for us to do _now_ - though this will become an issue in the future and I've filed #555 to discuss that)

**Comment by tm1000** [comment]
@shish thanks for all the work on this

FYI for anyone wanting to test against packages that are requiring 2.x you can simply add: `"thecodingmachine/safe": "3.0.0-alpha1 as 2.5.0"` to your own composer.json as a stopgap

**Comment by shish** [comment]
I'd have separate issues so it's easier to avoid any problems falling
between the cracks in the stream of comments :)

**Comment by michalschroeder** [comment]
Hey, I'm not sure if `phpstan-safe-rule:v1.3.1` is fully compatible with Safe v3. As I understood it correctly this [src/Utils/FunctionListLoader.php:fetchIndexedFunctions()](https://github.com/thecodingmachine/phpstan-safe-rule/blob/master/src/Utils/FunctionListLoader.php#L24C29-L24C50) function should be adjusted to support new directory structure?

**Comment by OskarStark** [comment]
> PS I wanted to push a branch with a fix but looks like I don't have proper permissions.

You should be able to push to your fork and create a PR against this repo afterwards

**Comment by shish** [comment]
(Also I published v1.3.1 a little early with _only_ the change of accepting safe v3, as I discovered a few other larger features I wanted to add to the rule, and I didn't want those changes to possibly cause disruption)

FWIW phpstan-safe-rule v1.3.1 and safe 3.0.0-beta2 are working together for me with composer.json like:
 
```jsonc
    "minimum-stability": "dev",  // to allow 3.0.0-beta2 when we ask for 3.0
    "require": {
        "thecodingmachine/safe": "^3.0",
    },
    "require-dev": {
        "phpstan/phpstan": "^2.0",
        "thecodingmachine/phpstan-safe-rule": "^v1.3.1"
    },

```

**Comment by michalschroeder** [comment]
Thanks, I've just updated project with:
- `composer require "thecodingmachine/safe:3.0.0-beta2 as 2.5.0"`
- `composer require thecodingmachine/phpstan-safe-rule:^1.3.1 --dev` 

Phpstan check worked now and it shows some errors like

```
Parameter #1 $array of static method Webmozart\Assert\Assert::count() expects array|Countable, array<string>|null given.
```

for the following code
```php
preg_match('/^\+(\d+)$/', $result, $matches);

Assert::count($matches, 2);
```
Is this expected? That `$matches` can be null instead of an empty array?

**Comment by shish** [comment]
> Is this expected? That $matches can be null instead of an empty array?

Yes, because phpstan doesn't have any magic for `Safe\preg_match` so it treats it like any regular function, and the type hint for that function's parameter is `string[]|null`

When Safe v3 is out, the next thing on my to-do list is looking at the way phpstan has special-cases for `preg_match`, copying that behaviour over to `Safe\preg_match`, and publishing 1.4.0 with the added magic (and dropped support for php 7.X)

**Comment by shish** [comment]
Safe v3.0 (and phpstan-safe-rule v1.4) published -- full changelogs below -- if you find problems in either, please report them as new issues rather than having them get lost in this thread, and we can get them fixed in point releases ❤️

https://github.com/thecodingmachine/safe/releases/tag/v3.0.0

https://github.com/thecodingmachine/phpstan-safe-rule/releases/tag/v1.4.0

**Comment by OskarStark** [comment]
Thanks for the release 🥰
